### PR TITLE
Article on API layers

### PR DESCRIPTION
**Pull request by jskeet**

### CODE REVIEW

File: docs/articles/api-layers.md [hunk 1]
```diff
@@ -1,3 +1,130 @@
...
+- [DatastoreClient](../obj/api/Google.Datastore.V1Beta3.DatastoreClient.yml) is generated by the
+  Google API Tools. This is still a higher level of abstraction than the raw RPC, supporting 
+  features such as [page streaming](page-streaming.md) and automatic creation of requests from the 
+  constituent parameters, but it does not provide the level of idiomatic usage that the hand-written
+  abstraction does. This is available from a `DatastoreDb` via the
+  [DatastoreDb.Client](../obj/api/Google.Datastore.V1Beta3.DatastoreDb.yml#Google_Datastore_V1Beta3_DatastoreDb_Client) property if necessary.
+- [Datastore.DatastoreClient](../obj/api/Google.Datastore.V1Beta3.Datastore.DatastoreClient.yml) is
+  the gRPC-generated RPC access class. Very few applications will need to use this, but it is available from the higher level `DatastoreClient` class via the
+  [DatastoreClient.GrpcClient](../obj/api/Google.Datastore.V1Beta3.DatastoreClient.yml#Google_Datastore_V1Beta3_DatastoreClient_GrpcClient) property if necessary.
+- Additional hand-written code in the form of extension methods and addition to the partial
```
Comment (chrisdunelm) [inline docs/articles/api-layers.md]: I wonder if this bullet should come second in this list? 
This list is from higher-level to lower-level code, except for this last point.

**Comment by chrisdunelm** [comment]
LGTM

File: docs/articles/api-layers.md (hunk 1 above)
Comment (jskeet) [inline docs/articles/api-layers.md]: I wondered about that. Maybe I should actually refactor it into "The library contains the following abstraction layers" and then have that last part as a separate paragraph. What do you think?

Comment (chrisdunelm) [inline docs/articles/api-layers.md]: Yes, I think that would be clearer.

**Comment by jskeet** [comment]
Performed some rewording - PTAL. (Will squash afterwards.)

**Comment by chrisdunelm** [comment]
LGTM

**Pull request merged**